drivers/gpu/drm/meson/meson_gem_prime.c

```diff
@@ -0,0 +1,239 @@
+
+	size = roundup(size, PAGE_SIZE);
+
+	scattered_obj = meson_drm_gem_init(dev, size);
+	if (IS_ERR(scattered_obj))
+		return scattered_obj;
+
+	scattered_obj->flags = flags;
+
+	ret = meson_drm_gem_alloc_buf(scattered_obj);
+	if (ret < 0) {
+		drm_gem_free_mmap_offset(&scattered_obj->base);
+		drm_gem_object_release(&scattered_obj->base);
+		kfree(scattered_obj);
+		return ERR_PTR(ret);
+	}
+
+	return scattered_obj;
+}
+
+void meson_drm_gem_destroy(struct meson_drm_gem_scattered_object *scattered_obj)
+{
+	struct drm_gem_object *obj = &scattered_obj->base;
+
+	drm_gem_free_mmap_offset(obj);
+
+	DRM_DEBUG_KMS("handle count = %d\n", obj->handle_count);
+
+	// TODO: we do not deal yet with dma_buf imported as a (scattered) gem object
+	if (!obj->import_attach) {
+		meson_drm_gem_free(scattered_obj, scattered_obj->nr_pages);
+		sg_free_table(scattered_obj->sgt);
+		kfree(scattered_obj->sgt);
+	}
+
+	drm_gem_object_release(obj);
+
+	kfree(scattered_obj);
+}
+
+void meson_drm_gem_scattered_free_object(struct drm_gem_object *gem_obj)
+{
+	meson_drm_gem_destroy(to_meson_drm_gem_scattered_obj(gem_obj));
+}
+
+struct sg_table *meson_drm_gem_scattered_prime_get_sg_table(struct drm_gem_object *gem_obj)
+{
+	struct meson_drm_gem_scattered_object *scattered_obj;
+
+	scattered_obj = to_meson_drm_gem_scattered_obj(gem_obj);
+
+	/* the returned sgt will be deallocated by the buffer-user when detaching */
+	return drm_prime_pages_to_sg(scattered_obj->pages, scattered_obj->nr_pages);
+}
+
+static int meson_drm_gem_handle_create(struct drm_gem_object *obj,
+				       struct drm_file *file_priv,
+				       unsigned int *handle)
+{
+	int ret;
+
+	ret = drm_gem_handle_create(file_priv, obj, handle);
+	if (ret)
+		return ret;
+
+	DRM_DEBUG_KMS("gem handle = 0x%x\n", *handle);
+
+	drm_gem_object_unreference_unlocked(obj);
+
+	return 0;
+}
+
+struct meson_drm_gem_scattered_object *meson_drm_gem_scattered_create_with_handle(
+		struct drm_device *dev, void *data,
+		struct drm_file *file_priv,
+		struct dma_attrs *dma_attrs)
+{
+	struct drm_meson_gem_create_with_ump *args = data;
+	struct meson_drm_gem_scattered_object *scattered_obj;
+	int ret;
+
+	scattered_obj = meson_drm_gem_create(dev, args->flags, args->size);
+	if (IS_ERR(scattered_obj))
+		return scattered_obj;
+
+	scattered_obj->dma_attrs = *dma_attrs;
+	scattered_obj->base.is_scattered = true;
+
+	ret = meson_drm_gem_handle_create(&scattered_obj->base, file_priv,
+					  &args->handle);
+	if (ret) {
+		meson_drm_gem_destroy(scattered_obj);
+		return ERR_PTR(ret);
+	}
+
+	return scattered_obj;
+}
+
+int meson_drm_gem_scattered_mmap_obj(struct drm_gem_object *obj,
+				     struct vm_area_struct *vma)
+{
+	struct meson_drm_gem_scattered_object *meson_gem;
+	unsigned long addr = vma->vm_start;
+	int i, ret;
+
+	meson_gem = to_meson_drm_gem_scattered_obj(obj);
+
+	for (i = 0; i < meson_gem->nr_pages; i++) {
+		struct page *page = meson_gem->pages[i];
+
+		ret = vm_insert_pfn(vma, addr, page_to_pfn(page));
```

Comment:
Because it is really expensive. I went down this road following this https://github.com/endlessm/linux-meson/blob/master/drivers/amlogic/gpu/mali/linux/mali_memory_os_alloc.c#L352